src/charm.py

```diff
@@ -2,203 +2,144 @@
+                    "override": "replace",
+                    "summary": "Spark Operator layer",
+                    "startup": "enabled",
+                    "command": (
+                        f"/usr/bin/tini -s -- /usr/bin/spark-operator -v=2 "
+                        "-logtostderr "
+                        f"-namespace={self.model.name} "
+                        "-enable-ui-service=true "
+                        "-controller-threads=10 "
+                        "-resync-interval=30 "
+                        "-enable-batch-scheduler=false "
+                        "-enable-metrics=true "
+                        "-metrics-labels=app_type "
+                        f"-metrics-port={self.model.config['metrics-port']} "
+                        "-metrics-endpoint=/metrics "
+                        "-enable-resource-quota-enforcement=false "
+                        "-enable-webhook=true "
+                        f"-webhook-svc-namespace={self.model.name} "
+                        f"-webhook-port={self.model.config['webhook-port']} "
+                        f"-webhook-svc-name={self.model.app.name} "
+                        f"-webhook-config-name={self._mutating_webhook_name} "
+                        f"-webhook-namespace-selector=model.juju.is/name={self.model.name} "
+                        "-webhook-fail-on-error=true"
+                    ),
                 }
             },
+        }
+        return pebble_layer
+
+    def _on_spark_pebble_ready(self, _):
+        self.unit.status = MaintenanceStatus("Configuring Spark Charm")
+
+        self.container.push("/etc/webhook-certs/ca-cert.pem", self._stored.ca, make_dirs=True)
+        self.container.push(
+            "/etc/webhook-certs/server-cert.pem", self._stored.cert, make_dirs=True
         )
-        self.model.unit.status = ActiveStatus()
+        self.container.push("/etc/webhook-certs/server-key.pem", self._stored.key, make_dirs=True)
+
+        self.resource_handler.apply()
+
+        self.container.add_layer(self._container_name, self._pebble_layer, combine=True)
+        self.container.autostart()
+
+        self.unit.status = ActiveStatus()
+
+    def _on_config_changed(self, event):
+        # check connection; config change could come before pebble ready
+        if not self.container.can_connect():
+            self.unit.status = WaitingStatus("Waiting to connect to spark container")
+            event.defer()
+            return
+
+        self.unit.status = MaintenanceStatus("Updating Spark Charm's Configurations")
+        self.service_patcher._patch(self.service_patcher)
+
+        self.container.add_layer(self._container_name, self._pebble_layer, combine=True)
+        self.container.replan()
+
+        self.unit.status = ActiveStatus()
+
+    def _on_remove(self, _):
+        manifests = self.resource_handler.render_manifests(force_recompute=False)
+        resources = apply_many(self.lightkube_client, manifests)
```

Comment:
I am a bit confused here, why are we applying and then removing? Is it to avoid the boilerplate that comes with getting lightkube rendered objects?

Reply:
yes, `apply_many` returns the resources that I can easily put into `delete_many`. I am still not very familiar with chisme or lightkube. I am open to suggestions on this

Reply:
It's fine, it is a limitation of chisme, we planned to add something to solve this, but didn't actually implemented it.

Reply:
I had expected this to work:
```python
manifests = self.resource_handler.render_manifests(force_recompute=False)
delete_many(self.lightkube_client, manifests)
```
Does render_manifests return a different form than delete_many accepts?  If so, we should probably change that

Reply:
+1 on @ca-scribner 's option

Reply:
That worked. Thanks.